### DISCUSSION TITLE
What's the purpose of BlockHandle and BufferHandle

### DISCUSSION THREAD

**Discussion by YjyJeff** [duckdb/duckdb#4597] [Q&A] [answered]
I am reading the source code of the BufferManager to learn how memory management works in duckdb. There are few docs in the source code, I can understand `FileBuffer`/`Block`/`ManagedBuffer`, but confused about the `BlockHandle` and `BufferHandle`. Why do these two classes exist? What's the relationship between all of these classes? Can someone explain it in detail? 

Thanks in advance~

**Comment by lnkuiper** [comment, marked as answer]
@YjyJeff 

Let's say that during execution we want a 1MB block, which can be offloaded to disk. We do:
```c++
auto block_handle = buffer_manager.RegisterMemory(1048576, false);
```

We get a `shared_ptr<BlockHandle>`. As long as there exists at least 1 `shared_ptr` to this specific `BlockHandle`, it is guaranteed that our block exists somewhere, either in-memory or on-disk.

Then, we want to use our block of memory, so we do:
```c++
auto buffer_handle = buffer_manager.Pin(block_handle);
```

We get a `BufferHandle`. As long as at least one `BufferHandle` for this specific block exists, the block is guaranteed to be in-memory, so then we can get a pointer to the start of 1MB block like so:
```c++
auto block_ptr = buffer_handle.Ptr();
```

So, the general distinction is that `BlockHandle` manages the 'lifespan' of the block, whereas `BufferHandle` is used to directly access the data in the block.

**Reply by YjyJeff** [reply]
Awesome, great design~
Could you answer the other question about memory management? 

I find that all of the `ManagedBuffer` whose size equals `BlockSize` are managed by `TemporaryFileManager`. It works as: a temporary file stores multiple `ManagedBuffer`, rather than one file for one buffer. However, the `ManagedBuffer` whose size is larger than `BlockSize` are stored in each file, one file for one buffer. 

What's the advantage of this design?  Is it better than one file for one buffer for all `ManagedBuffer`?

Thanks in advance~
